### PR TITLE
Add power boost sensors

### PR DESCRIPTION
**Pull request by martonbtoth**
This commit adds support for the [Wallbox Power Boost feature](https://support.wallbox.com/na/knowledge-base/energy-management-solutions/#powerboost). This is basically a separate device (seems like its a rebranded Inepro PRO380?), which is connected to the charger over modbus. The stock firmware conveniently puts a bunch of statistics about the device in Redis, which this PR exposes as sensors for Home Assistant. It basically makes it possible to monitor the energy usage of a whole household through this feature.

I hope you like it, I threw this together in less than an hour and I actually cannot test it without a power boost device.

### CODE REVIEW

**Comment by jan--f** [comment]
Nice! I proposed exactly that in https://github.com/jagheterfredrik/wallbox-mqtt-bridge/discussions/48. You beat me to it by a weekend :) 
Going by the discussion it might be worth it to put this behind a config field or cli flag for users that don't want or need this device.

File: app/sensors.go [hunk 1]
```diff
@@ -162,6 +162,95 @@ func getEntities(w *wallbox.Wallbox) map[string]Entity {
 				"suggested_display_precision": "1",
 			},
 		},
+		"power_boost_power_l1": {
+			Component: "sensor",
+			Getter: func() string {
+				return fmt.Sprint(w.Data.RedisM2W.PowerBoostLine1Power)
+			},
+			Config: map[string]string{
+				"name":                        "Power Boost L1",
```
Comment (jan--f) [inline app/sensors.go:171]: Minor nit: I'd rename `Power Boost` to something more generic like `External Energy Meter` or similar.

Comment (martonbtoth) [reply app/sensors.go:171]: I agree that the name is not ideal, but it seems like Wallbox also supports [MID type energy meters](https://support.wallbox.com/na/knowledge-base/accessory/#power-meter), so I think anything else would be more confusing if somebody adds support for those. I don't like calling it power boost either, so I'm open to ideas though. :)

**Comment by martonbtoth** [comment]
> Nice! I proposed exactly that in #48. You beat me to it by a weekend :) Going by the discussion it might be worth it to put this behind a config field or cli flag for users that don't want or need this device.

I think it should be possible to read this config from the device directly, the `power_management_config` table contains a `power_boost_enable` column, which is currently `1` for me. Could somebody please check without power boost enabled? I'm a bit afraid to disable it.

**Comment by jagheterfredrik** [comment]
```
MariaDB [wallbox]> select power_boost_enable from power_management_config;
+--------------------+
| power_boost_enable |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)
```

**Comment by tronikos** [comment]
Please allow this to be disabled even if power_boost_enable is 1. See the previously linked discussion for why.

**Comment by martonbtoth** [comment]
@tronikos Based on the linked discussion I understand that you are concerned about the resource utilization this feature would add. 

I think the additional CPU cycles required is so negligible that you won't see a difference with the feature enabled vs disabled: it does not do additional round trips to Redis and is already heavily rate limited. CPU usage is so low in fact, that even `top` consumes more.

Based on that I would advise putting every new feature behind a config entry, as it is going to just add code complexity and increases friction for the users.

That being said, I'm happy to add code to handle the config entry and also check if Power Boost is enabled on the charger, if that's what it takes to merge this feature. What do you guys think?

**Comment by jagheterfredrik** [comment]
I think "Power Boost" naming makes sense as that's what Wallbox uses. I think I agree that it'd be nice to add this as a config (config.go, tui.go), default to false (to preserve functionality for existing users).

File: app/sensors.go [hunk 2]
```diff
@@ -162,6 +163,116 @@ func getEntities(w *wallbox.Wallbox) map[string]Entity {
...
+				return config.Settings.PowerBoostEnabled
+			},
+		},
+		"power_boost_current_l2": {
+			Component: "sensor",
+			Getter: func() string {
+				return fmt.Sprint(w.Data.RedisM2W.PowerBoostLine2Current)
+			},
+			Config: map[string]string{
+				"name":                        "Power Boost current L1",
```
Comment (jan--f) [inline app/sensors.go:236]:
```suggestion
				"name":                        "Power Boost current L2",
```

File: app/sensors.go [hunk 3]
```diff
@@ -162,6 +163,116 @@ func getEntities(w *wallbox.Wallbox) map[string]Entity {
...
+				return config.Settings.PowerBoostEnabled
+			},
+		},
+		"power_boost_current_l3": {
+			Component: "sensor",
+			Getter: func() string {
+				return fmt.Sprint(w.Data.RedisM2W.PowerBoostLine3Current)
+			},
+			Config: map[string]string{
+				"name":                        "Power Boost current L1",
```
Comment (jan--f) [inline app/sensors.go:252]:
```suggestion
				"name":                        "Power Boost current L3",
```

**Comment by martonbtoth** [comment]
Added the extra configuration field, I hope you guys like it! I also think it would be beneficial to start documenting the config file, but thats for another PR! :)

**Comment by yavinmaster** [comment]
I have the CT clamp version of Power Boost.  Are the same parameters read for this version as well as the full power meter version or do we need two versions of the config?

**Comment by martonbtoth** [comment]
@yavinmaster Since @jan--f and I have different power boost devices and they both have the generic `PBO` keys in redis, I think it should work with your hardware too. Would you be able to check if you have [those keys](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/pull/50/files#diff-307cf37945e5a2e6ce76176c3217a0d1cffb6ef40e629fdce8a303b54e32f30eR40) set on your wallbox's redis?

**Comment by jagheterfredrik** [comment]
Preliminary looks good, will have a proper look tonight

**Comment by jagheterfredrik** [comment]
I'd definitely prefer if the entities followed the pattern used by [debugEntites](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/blob/9da0c342387642d9dc9f25c405e30c8c715990d9/app/sensors.go#L226). I.e. check the config in [the bridge](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/blob/9da0c342387642d9dc9f25c405e30c8c715990d9/app/bridge.go#L28) and then call e.g. `func getPowerBoostEntities(w *wallbox.Wallbox, c *wallbox.WallboxConfig) map[string]Entity`.

**Comment by martonbtoth** [comment]
Sure, no problem! I'll try to get it done this evening! 👍

**Comment by jagheterfredrik** [comment]
Thanks @martonbtoth 👍

**Comment by yavinmaster** [comment]
> @yavinmaster Since @jan--f and I have different power boost devices and they both have the generic `PBO` keys in redis, I think it should work with your hardware too. Would you be able to check if you have [those keys](https://github.com/jagheterfredrik/wallbox-mqtt-bridge/pull/50/files#diff-307cf37945e5a2e6ce76176c3217a0d1cffb6ef40e629fdce8a303b54e32f30eR40) set on your wallbox's redis?

Yes - they appear to be there.  Obviously having only a CT, I only have a current value.

**Comment by CaptInsano** [comment]
Thanks so much for adding this feature.

I just got a chance to install this onto my wallbox pulsar plus with power boost today. 

I am unsure of the hardware used in my install for monitoring my house energy usage but is it normal for my sensors to be as follows?:

WhatsApp Image 2024-04-10 at 17 22 58

I understand that L2 and L3 entities are blank as I am a single phase install but should "Power Boost L1" be giving a value in watts or should it be empty?

I am trying to add it to my HomeAssistant instance to enable whole house energy monitoring but I am unsure if I need a wattage input?

Thanks again for all the work on this project by all.

**Comment by jan--f** [comment]
That depends on the external power meter. There are models that only supply current values but no power.

**Pull request merged**